### DISCUSSION TITLE
Localization not working

### DISCUSSION THREAD

**Discussion by juagicre** [oleg-shilo/wixsharp#1438] [Q&A] [answered]
I'm checking the current documentation in this repo to localize the .msi setup. Until now all my efforts were futile.

In the first image I can [see](https://github.com/oleg-shilo/wixsharp/wiki/Localization#msi-localization) there is already some differences between what I see in the image, the UIText entry, which does not appear on my .msi even I do include the WixUI_de-DE.wxl (that comes from the repo as well). I do this:

    // #1
    project.Language = "en-US";
    var productMsi = project.BuildMsi();

    // #2-3
    project.Language = "de-DE";
    project.LocalizationFile = "Resources\\Locales\\WixUI_de-DE.wxl";
    string mstFile = project.BuildLanguageTransform(productMsi, "de-DE");

    // #4
    productMsi.EmbedTransform(mstFile);

    // #5
    productMsi.SetPackageLanguages("en-US,de-DE".ToLcidList());

Those are the only translations I see on my de-DE.msi file:
![image](https://github.com/oleg-shilo/wixsharp/assets/1616825/b7fa1a08-1667-434f-988b-674d959ea0e2)
But nothing on the original setup.msi file which should be the combination of both.
Also I don't find any property named: WixUIBack as I see in the wxl and in the WPF dialog.

I guess I'm missing something big here... anybody to throw some ideas or tips?

Thanks in advance!

**Comment by oleg-shilo** [comment]
Did you have a look at this sample? 
![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/a697685d-394d-4b47-9706-83b78a0ffe02)

It might be exactly what you are looking for.
![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/6a5d44d6-1b11-4de9-920d-9dfe21f589a6)

It seems to work exactly as expected.

**Reply by juagicre** [reply]
I saw this example, yes, but as I'm at the moment not using bootstrapper, at least intentionally, I was afraid this could make things more complicated... I'll have a look at it again and try to follow it :)

**Comment by oleg-shilo** [comment, marked as answer]
In the sample, it uses bootstrapper only to let user select the desired language and then run the msi with the appropriate CLI param.
Otherwise, you will meed to ask user not to double-click msi but run it from the terminal as an extra CLI parameter will be needed depending on the language selection.

Thus you may want to automate all this via the bootstrapper app. BTW MSI does not support authoring multi-language MSIs. The bootstrapper sample only shows the MSI-transformation trick MSI uster started using for multilanguage scenarios.

Alternative, you can look at pure WixSharp multilanguage support. No bootstrapper is required. Pure single MSI with two `*.wxl` embedded.

Thin this is the example for you:
![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/68be35fb-e1cd-43c6-a379-eec287350d9c)

**Reply by juagicre** [reply]
I like the idea of embedding all the needed *.wxl! (and avoid the use of bootstrapper is a very big plus here)
In my scenario the decision should be automatic one, based on OS current culture... let's see if I can manage to get some inspiration from your solution :)
Thanks a lot!!

**Comment by juagicre** [comment]
I could get it working :) 
A bonus question... I need to add a few new entries to those files, for this reason I guess if I just add as binary the updated english (WixUI_en-US.wxl) with the new entries (also new entries will be in the translated files, of course) that will override the default one?

**Reply by juagicre** [reply]
I answer myself, yes, it works with new entries adding the default language file as well.

**Comment by juagicre** [comment]
A new question related to all this... I want to show a translatable message from a custom action message dialog... it looks like this:

    session.Message(
    InstallMessage.Error,
    new Record("[UICustomTextAppStillRunning]"));
    //original string now in the wxl:
    //new Record("App is running. Please save and close and run the setup again."));

I see in the localization documentation this code:

    MsiRuntime runtime = e.ManagedUI.Shell.MsiRuntime();
    if (CultureInfo.CurrentUICulture.TwoLetterISOLanguageName == "de")
        runtime.UIText["Copy"] = "Kopieren";

Maybe I could get it like this:

    myString = runtime.UIText["UICustomTextAppStillRunning"]

but I don't know how to get the runtime from the CustomAction.... I just have the Session, but I don't have the "e":

    MsiRuntime runtime = e.ManagedUI.Shell.MsiRuntime();

**Comment by oleg-shilo** [comment]
The `Session` object gives you access to all localization data that you need. Either MSI localization (UIText table) or WixSharp localization ( wxl file in Binary table).  

`MsiRuntime` type is a class for easy access to the session data and in the case of `ManagedProject` it is easily available with the session event argument. However, if you choose to go with low-level session events (raw Custom Actions) then you need either to access session data manually or construct `MsiRuntime` onfly:
 
```C# 
var runtime = new MsiRuntime(session);
MessageBox.Show("Click [WixUINext]".LocalizeWith(runtime.Localize));
```

Note, `WixUINext` is a string from the localizable UI content.

I reflected our conversation here: https://github.com/oleg-shilo/wixsharp/wiki/Localization#custom-content-localization

**Comment by juagicre** [comment]
I tried your code and works fine on tags that are already defined the the wxl file, but new ones I added at the end they are not being found.
So this code is inside my custom action, in case I return ActionResult.Failure I do show the message:

    var runtime = new MsiRuntime(session); // session comes as an argument from the CustomAction method.

    var msg = "Custom msg: [UICustomTextAppStillRunning] translated!";
    var rawMsg = "Std msg: [AdvancedWelcomeEulaDlgLicenseAcceptedCheckBox] translated!";

    var translatedMsg = msg.LocalizeWith(runtime.Localize);
    var translatedRawMsg = rawMsg.LocalizeWith(runtime.Localize);

    MessageBox.Show(translatedMsg);
    MessageBox.Show(translatedRawMsg);

    //session.Message(InstallMessage.Error, new Record(translatedMsg));

And the outcome is:
First message box:

> Custom msg: UICustomTextAppStillRunning translated!

Second message box:

> Std msg: I accept the terms in the License Agreement translated!

So I suspect that for whatsoever reason my new entry in the wxl is not present, parsed... here how it is defined in the wxl:

    <String Id="UICustomTextPconProStillRunning" Overridable="yes">APP is running. Please close it before running the setup.</String>

**Comment by oleg-shilo** [comment]
I do not know what custom action you are invoking your code from. 

I gave you an example of ManagedUI. I tested it and it works, unless you detected its failure. Not sure if you try to run it?

Without the context, it's hard to say if you run the same runtime initialization as in my sample or the MSI has the required wxl embedded.

Quite possibly you are not building your setup as a `ManagedProject` but as `Project`. If this is the case then MSiRuntime will not be initialized the same way and one extra step might be required.


```C#
// WixSharp.ManagedProject
static void Project_UIInitialized(SetupEventArgs e)
{
    var runtime = new MsiRuntime(e.Session);

// WixSharp.Project
[CustomAction]
public static ActionResult InvokeRemoveFiles(Session session)
{
    var runtime = new MsiRuntime(e.Session);
    runtime.UIText.InitFromWxl(e.Session.ReadBinary("exta.wxl"), merge: true);
 ```

If this does not help, please compose a simple HelloWorld style project that shows the problem and I will have a look at it.

**Reply by juagicre** [reply]
I did check and I use a `ManagedProject` as well, so your example should be working, but for some reason unknown to me it doesn't.
There are two issues in the `CustomAction`, first one is that the `standard translation` tag is not translated and the second one that the `new custom translation` tag is not recognized. 

I uploaded a [sample project](https://github.com/juagicre/WixSharp.HelloWorldWpfTranslations) to github demonstrating the behavior I'm facing in my project. 

It enforces the "es" wxl to be used. Not all sentences are translated but most of the ones that appear, and specially the ones I want to be translated for the example :) 
Note that you need to run notepad++ to get the CustomAction message to appear!
Not yet sure if this could be a bug, most probably I'm defining things in the wrong way...

**Comment by oleg-shilo** [comment]
You are doing everything right. But...
You became a victim of a nasty MSI runtime hosting model. I've been there too. :)

The initialization of the localization routine (`runtime.UIText.InitFromWxl...`) is done within the boundaries of the CA it's invoked.
Thus in the UIInitializaed even handler, the localization mapping will be active until this CA is active (UI dialogs are displayed). 
But in your case, you are trying to access this data from another CA that has not been initialized with your xsl file. Why? Wait for it....
Msi runtime started a brand new rendll32.exe process for every single CA. Thus your previous call `InitFromWxl` has no effect.

You can fix it by simply calling `InitFromWxl` from your custom action. Or by invoking your `CheckProcesses` from the `UIInitialized` event handle, just after you call `InitFromWxl`:

![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/15037321-e7d0-42db-b460-08f0f656cb30)

**Reply by juagicre** [reply]
I see the two options as quite valid and I can confirm the code you shared for the first option works fine :) 

Probably in this case the best approach would be your second mentioned option: to invoke the CheckProcesses from the project.UIIInitialized defined in the Localize static method. So the message is shown at the very beginning and setup is stop earlier. 

Unfortunately I could not get to work... I did comment this (at the Program.cs Main class) not to have the CA check as I used to have:

```
project.Actions = new[]
    {
    new ManagedAction(
        CustomActions.CheckProcesses,
        Return.check,
        When.Before,
        Step.InstallInitialize,
        Condition.NOT_Installed)
    };
```

And at the UIInitialized I did add the call to the CA like this: 

```
project.UIInitialized += (SetupEventArgs e) =>
            {
                MsiRuntime runtime = e.ManagedUI.Shell.MsiRuntime();

                CultureInfo culture = CultureInfo.InstalledUICulture;
                //var osLanguage = culture.TwoLetterISOLanguageName.ToLower();
                var osLanguage = "es";
                switch (osLanguage)
                {
                    case "de":
                        runtime.UIText.InitFromWxl(e.Session.ReadBinary("de_xsl"), merge: true);
                        break;
                    case "fr":
                        runtime.UIText.InitFromWxl(e.Session.ReadBinary("fr_xsl"), merge: true);
                        break;
                    case "es":
                        runtime.UIText.InitFromWxl(e.Session.ReadBinary("es_xsl"), merge: true);
                        break;
                    default:
                        runtime.UIText.InitFromWxl(e.Session.ReadBinary("en_xsl"), merge: true);
                        break;
                }

                // Call CA to check for planner process:
                // runtime.MsiSession // not translated
                // e.Session // not translated
                // runtime.Session // it's an ISession that does not cast to Session
                //Session mySession = (Session)runtime.Session.SessionContext;
                
                CustomActions.CheckProcesses(mySession);

            };
```

Probably the main problem is that I'm using the wrong session, but I tried with all the ones you see commented and I always got no translation. 

Also in the return I don't know how to "abort" in a clean way the installation, should I check for the ActionResult and then, how to say to the installer that it should show the last UI dialog with the default message that says that something went wrong? Not sure if closing/exiting the installation would cause any problem.

Sorry for making this thread so long! Maybe it looks a bit OT but I still see that this complete flow (custom translation+CA+exit in a clean way) is of quite big interest to the basic setup makers, so having all this in one thread looks to me still makes sense.

**Reply by oleg-shilo** [reply]
Not a prob. I can create a working sample of that.

Do I understand the requirements right that you need:
- to add 4 language files to setup before showing the GUI
- to popup the localized message box 
- then abort the setup. 
 
Am I right?

**Comment by juagicre** [comment]
Yes, for now it is 4 languages, and yes, aborting the setup in a clean way (not messing anything related to running again the setup).Thanks in advance for all the support!

**Comment by oleg-shilo** [comment]
Done. With the attached sample the checking is done before the UI is displayed.
[WixSharp.HelloWorldWpfTranslations.zip](https://github.com/oleg-shilo/wixsharp/files/14275141/WixSharp.HelloWorldWpfTranslations.zip)

**Reply by juagicre** [reply]
That works great! Thanks a lot for your help and support!
